### ISSUE TITLE
Lots of 'timeout of 48000ms exceeded' during the day, but the site is still up?

### ISSUE THREAD

**Issue by Oaktribe** [louislam/uptime-kuma#275]
**Is it a duplicate question?**
Do not thing so.

**Describe the bug**
During the day numerous alerts of a monitor timing out with the message 'timeout of 48000ms exceeded' is reported. Then it works again, and then another timeout message is reported.

I setup a custom .js script checking the site the same way Kuma does outside of the docker, and nothing is reported there during the day.
Could it be some issue with Docker for Desktop on Windows 10 with WSL2 producing false timeouts?

**To Reproduce**
Add monitor to site, and wait. That's how I do it.

**Expected behavior**
Not to throw a timeout since the site does respond.

**Info**
- Uptime Kuma Version: 1.3.2
- Using Docker?: Yes
- OS: Windows
- Browser: Firefox

**Screenshots**
![1](https://user-images.githubusercontent.com/89459766/130967066-fd711a78-26e8-4f61-8e57-5899742fa3a9.png)

**Error Log**
Nothing is logged to the docker log window.

**Comment by Oaktribe** [comment]
Did some Googling and found this.
https://github.com/microsoft/Windows-Containers/issues/145

So I ran that command on my Windows Host machine, restarted and fired up the docker again. So far it seems to have helped, before there would be at least 10 timeouts a day. It's been running for 48 hours now without any timeouts.

**Comment by agrider** [comment]
I'm experiencing the same issue, but mine is running in Docker on Ubuntu 20.04.3. Like OP, the performance/response times that kuma is reporting it's getting aren't consistent with the actual response times of my services. I get an alternating graph of timeouts and outlandish response times like 30-40s when the actual service is responding in <3s consistently outside of Kuma. Wondering if I should open a new issue or if this one needs re-opened since it's not just specific to Docker in Windows. @louislam preferences?

**Comment by louislam** [comment]
@Oaktribe @agrider Because I cannot reproduce the problem, it really hard to address the problem.

However, one of our contributions discovered that it could be Alpine Docker problem.
https://github.com/louislam/uptime-kuma/issues/294#issuecomment-909353979

I will build a Debian/Ubuntu docker later for you guys testing.

@agrider 
If possible, try to run it without Docker:
https://github.com/louislam/uptime-kuma/wiki/%F0%9F%94%A7-How-to-Install#-without-docker-recommended-for-x86x64-only

**Comment by agrider** [comment]
@louislam I don't install basically anything bare metal on my server if I can help it, so I can't help with the non-docker option, but I did clone down the repo to my Gitlab instance and rebuilt the image on Debian Bullseye instead to see if it's an Alpine issue. I'm going to leave it up and running a day or two and see what the result is.

**Comment by louislam** [comment]
> @louislam I don't install basically anything bare metal on my server if I can help it, so I can't help with the non-docker option, but I did clone down the repo to my Gitlab instance and rebuilt the image on Debian Bullseye instead to see if it's an Alpine issue. I'm going to leave it up and running a day or two and see what the result is.

Thank you so mcuh!

**Comment by agrider** [comment]
Ok, so I don't need to wait a day, I can pretty conclusively say that the issue is not alpine related as my times were still atrocious, which doesn't surprise me as I use numerous other alpine based images with no latency issues. I did resolve my issue though. When I stood up my instance in my compose file, I forgot to apply my default configs anchor, one of which is setting DNS directly on the container to point to my local unbound instance. As soon as I set DNS to go directly to the DNS server instead of routing through the docker gateway for DNS resolution (which still ultimately landed at unbound), my times dropped to the expected values of <100ms, under both the image I made and the latest alpine based image. My guess is that the if you have a large number of containers like I do (>80), or a high check rate set, that you can overwhelm the DNS resolver in the Docker gateway. Long story short, it's not a conclusive fix necessarily, but I'd try setting the DNS on your kuma container to your preferred upstream server directly instead of letting the docker gateway manage it.

That said, is uptime-kuma set to cache DNS results and respect the TTL's of DNS records it requests? I have the vast majority of my local DNS entries set to an enormously large TTL, so even with the DNS going through the docker gateway and a high check rate, there should have been a relatively small number of DNS queries, but it looks like that wasn't the case, based on the traffic I was seeing it seemed like Kuma was making a DNS request per check.

**Comment by louislam** [comment]
> I was seeing it seemed like Kuma was making a DNS request per check.

Thank you for your finding, it seems that it is worth to implement dns cache in Uptime Kuma. I always thought dns cache is managed by OS.

**Comment by agrider** [comment]
It usually is, though Alpine is a pretty sparse distro, so it may not have a caching resolver configured/enabled by default? I'm really not familiar enough and haven't encountered dns issues like this with it before.

It also might and my conjecture about root cause could be totally wrong, in which case I'm unsure why switching to direct dns config resolved the issue.

**Comment by yasharne** [comment]
same problem on `1.10.2`, switching to debian based image fixed the problem

**Comment by csakaszamok** [comment]
v1.21.3
unfortunately this false positive error still occurs every day

**Comment by UtechtDustin** [comment]
We have the same issue with Kuma Version: 1.21.3, Kuma is installed without docker.   
So the docker-image can't be the issue, it's running on our infrastructure server which also includes an DNS-Server.

**Comment by CommanderStorm** [comment]
@csakaszamok @UtechtDustin @yasharne 
Have you activeted the DNS-cache?
![image](https://github.com/louislam/uptime-kuma/assets/26258709/62646d6b-76dd-413d-83db-3427e51e71e2)

**Comment by csakaszamok** [comment]
It was disabled.
Now I've set to enable, thx the tip

**Comment by UtechtDustin** [comment]
It was disabled, lets see if that fixes it.

**Comment by UtechtDustin** [comment]
@CommanderStorm it seems that option don't fixed the issue.     
Last night we got a lot of "spam"-messages(32 Messages within a few Minutes) with the same errror `timeout of 48000ms`.    
I'm not sure why the DNS should be the cause of an timeout of 48 Seconds.

**Comment by solracsf** [comment]
https://github.com/louislam/uptime-kuma/pull/3472 could help here.
To be shipped on 1.23.0.

**Comment by Aj7Ay** [comment]
> @csakaszamok @UtechtDustin @yasharne Have you activeted the DNS-cache? ![image](https://private-user-images.githubusercontent.com/26258709/245832644-62646d6b-76dd-413d-83db-3427e51e71e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM3ODI1MDUsIm5iZiI6MTcxMzc4MjIwNSwicGF0aCI6Ii8yNjI1ODcwOS8yNDU4MzI2NDQtNjI2NDZkNmItNzZkZC00MTNkLTgzZGItMzQyN2U1MWU3MWUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIyVDEwMzY0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg1MjczMDFiMjRmZTY3YzE2MjY0NjkxNjIyYjI3YjFhMjllNTAxOTA1ODZjYjcwZjlhYzQwNzM2MTJkZjM2MmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Y5fo7Y6tLpC1jrr4pjZTwwvbmeFSot5em37nJ1K8ybc)

Its Deprecated any new option adding ?

**Comment by CommanderStorm** [comment]
What are you asking?

I think you are asking what we replaced this with.
We have replaced this with the [Name service caching daemon](https://de.wikipedia.org/wiki/Name_service_caching_daemon) if you are using the docker container. For a native installation, you will have to install this yourself.

**Comment by UPSOKen** [comment]
I am still experiencing this issue on Ubuntu docker. Majority of my sites with a specific hosting company do this ALL day, even though the sites are all working fine.

**Comment by af7567** [comment]
I am suddenly getting this issue today in Linux docker and can't get the monitor to go green again. The site is up and I can run curl from within the container to access the site OK.
I also tried deleting and recreating the container and changing all the DNS cache options.
It has been working fine for months and there have been no system changes/updates today which could have caused it.

**Comment by CommanderStorm** [comment]
@af7567 [Happy eyeballs v2](https://datatracker.ietf.org/doc/html/rfc8305) is a feature that node implemented [in a more recent version](https://github.com/nodejs/node/issues/41625). V2.0 has switched to said version.
We need a few things to be able to release that. See #4500 for further context.

**Comment by GordonHannan** [comment]
> I am still experiencing this issue on Ubuntu docker. Majority of my sites with a specific hosting company do this ALL day, even though the sites are all working fine.

You're not alone, I'm seeing it too.  Name service caching daemon is enabled and I've tried on the public release and the beta. Using Ubuntu with docker (using cloudflare tunnel though). Not sure if that is causing issues or not?

**Comment by puregraphx** [comment]
Could it help if you set up Retries to 1 or higher? I had 1 site on a Hetzner server that also hosts other WP sites. This one site got occasional 48s time outs, but it was the only site on Kuma without Retries.

**Comment by jlahijani** [comment]
I'm having this issue too.  I'm running on a cheap Hetzner VPS with Ubuntu 24.04, installed Uptime Kuma the normal way (not using Docker).

I'm running a version of Uptime Kuma from August 2024, but will update to latest to see if it helps.  If I don't follow up, it probably worked.

**Comment by robsontenorio** [comment]
@jlahijani The latest stable version 1.23.6 still has the same issues. Occasionally, I receive false positives indicating that the system is down, but I can immediately verify that it is online.

**Comment by leikoilja** [comment]
> I'm having this issue too. I'm running on a cheap Hetzner VPS with Ubuntu 24.04, installed Uptime Kuma the normal way (not using Docker).
> 
> I'm running a version of Uptime Kuma from August 2024, but will update to latest to see if it helps. If I don't follow up, it probably worked.

have the same issue running on docker (under ubuntu 24), the issue with the v1 and with 2.0.0-beta.3 as well. make it a bit hard to trust uptime kuma to monitor infra reliability :/  
Edit: as a temporary walk-around i will add +1 to retries before marking the service down, will see if that help hide false positives
![Image](https://github.com/user-attachments/assets/b42a4fdf-44cd-4c4a-bc47-a8708b87eb1a)

**Comment by raabmar** [comment]
Have you tested pausing Cloudflare if it's in use? I think it might be related to Cloudflare

**Comment by jorismak** [comment]
What is, the timeouts? Lots of people without any cloudflare setup that have it. And the cloudflare tunnel is (often) used to access the machine and the web-ui, not to let the machine make outside connections.

I do see a lot of people with the docker image having this issue.

Oh, I just realized... my vps also has cloudflared running to access it's SSH.

**Comment by raabmar** [comment]
Also on the docker image here. It's really annoying and I have seen this many times now in the last year or so.

I spent again a week on this now. It's so random. Most of the websites I monitor have an identical setup but out of nowhere those timeouts and false down alerts show on a single website. No change happened previously to that website. Not one single time could I actually see the website down by checking myself in the browser.

I migrated the website in question to a different server with a different hosting provider and it showed the same problem.

Either Uptime Kuma waits for Cloudflare and Cloudflare does not respond in time. Or Cloudflare waits for the origin server and the server does not respond in time. Maybe it's this combination that leads to too much waiting because without Cloudflare no problem at all. I tested Uptimerobot and it does not report a downtime or timeout one single time, and it has its default timeout set to 30 seconds which is even a lot lower.

That means for me ...
- Pause Cloudflare makes a difference → no more timeouts but not a solution as Cloudflare is a must have
- Whitelisting the IP on Cloudflare → does not make a difference but always worth a try
- Set Retry to 1 in Uptime Kuma should work because I have never seen 2 timeouts in a row → in my eyes a workaround not a solution
- Changing Heartbeat should not make much of a difference, only lead to fewer timeouts but the problem stays the same
- Uptimerobot does not show the same problem even though Cloudflare is enabled
- => doesn't that all point to Uptime Kuma itself?

I see those discussions are open for years. How come? Setting Retry to 1 is not the correct solution, it should never take > 48 seconds

**Comment by jorismak** [comment]
Its something memory related or intensive related. I had hundreds of tests,
and introduced the interval and gave them all a retry and it's way less. It
has nothing to do with Cloudflare (at least none of the sites i test are
Cloudflare related.).

Setting a memory limit on the socket container seemed to help a bit, but
not all.

Restarting the container automatically every day or so seems to help.

Note this is all with kuma1, I haven't tested the test versions for 2 yet.

**Comment by raabmar** [comment]
Same - everything I said was related to the latest Uptime Kuma 1.x and not the 2 (beta).

Maybe in your case it's a different source. I can clearly get rid those timeouts by just pausing Cloudflare. I am trying to read up now on Docker and DNS

**Comment by raabmar** [comment]
**Adding context:** I use the docker run command and not docker compose. I just upgraded from 1.23.13 to 1.23.16 as I was not on the latest stable version anymore. I modified the run command:
```
docker run -d --restart=always -p 127.0.0.1:3001:3001 --dns=1.1.1.1 --dns=8.8.8.8 -v uptime-kuma:/app/data --name uptime-kuma louislam/uptime-kuma:1

```

This is just like the official command — except:
- I bind to 127.0.0.1 (because you use Nginx proxy)
- I add --dns for reliable hostname resolution

All else is the same.

Now testing if it makes a difference in the monitorings when I switch back from Retry = 1 to Retry = 0 in the monitor.

**EDIT:** Update, it did not work. Maybe they are less, maybe not, but they downtime alerts / timeouts are still happening.

**Comment by sobakintech** [comment]
I have the same issue. Is there a way to fix it yet? I tried whitelisting my servers IP in Cloudflare, didn't work. It's only happening on Cloudflare protected websites for me

**Comment by robsontenorio** [comment]
So … 

1) If I run KumaUptime using Docker …
2) And my site is behind the Cloudflare proxy, then …
3) I will see that timeouts. 

At least we have a logic here 😂

I don’t know what else can be fixed here. Is it Cloudflare or KumaUptime/Docker fault ?

**Comment by jorismak** [comment]
Again, I say Cloudflare has nothing to do with it, or is a coincidence, or
a completely different issue of its repeatable.

This issue started - and still is a thing with me - when testing sites
without Cloudflare.

Ik actually hearing for someone who has these issues and is NOT using
docker to run Uptime Kuma. Because maybe that's the common ground we're
looking for.

**Comment by raabmar** [comment]
## Ongoing Timeout Issue in Uptime Kuma - Summary of 4 Years of Reports

I’ve tried to summarize the current state of this long-standing issue, referencing several contributors from this thread and beyond. I’ve been personally affected by it for over a year now and want to help narrow it down. I think there are more Github issues discussing the same or similar thing.

---

### 1. Cloudflare Appears to Be a Key Factor

@jorismak - Cloudflare has definitely something to do with it. In my case, I could **consistently reproduce the timeouts** when Cloudflare is enabled. Pausing Cloudflare immediately solved it. However, out of the approx. 50 monitors I have on WordPress websites it might show up only on one or two sites. It's all very random. My latest issue started June 2nd, before that there were no timeouts/downtimes at all for the particular site. In a few months the same might start with any other monitor/site. I have seen it like that too many times in the last 1+ year(s).

However, the fact that it **also happens without Cloudflare** suggests that Cloudflare might **amplify or trigger** an underlying problem, rather than being the sole cause.

---

### 2. Docker DNS Resolution vs. External Providers

I thought that using external DNS resolvers (Cloudflare & Google) in the Docker run command would help - it didn’t. @CallMeTerdFerguson was able to solve it in 2021 for his instance as you can see [here](https://github.com/louislam/uptime-kuma/issues/275#issuecomment-909502971), but that was a bit advanced for my level.

Here’s the command I used:

```bash
docker run -d --restart=always -p 127.0.0.1:3001:3001 \
  --dns=1.1.1.1 --dns=8.8.8.8 \
  -v uptime-kuma:/app/data \
  --name uptime-kuma louislam/uptime-kuma:1
```

Despite this, I still received **hourly timeouts**.

This suggests that even explicitly setting external DNS resolvers inside Docker is **not enough** to eliminate the problem. There might be other network stack issues at play within Docker, especially when combined with Cloudflare.

---

### 3. It's Not Only Docker - Native Installs Affected Too

@louislam (2021):

> If possible, try to run it without Docker  
> 👉 https://github.com/louislam/uptime-kuma/wiki/🔧-How-to-Install#-without-docker-recommended-for-x86x64-only

@UtechtDustin (2023):

> We have the same issue (multiple times per day) with Kuma Version: 1.21.3, Kuma is installed **without Docker**

@jlahijani (May 2025)

> I'm having this issue too. I'm running on a cheap Hetzner VPS with Ubuntu 24.04, installed Uptime Kuma the normal way (not using Docker).


This clearly indicates that **Docker is not the root cause**, although it may add complexity or latency. At least two people reported this within this thread.

---

### 4. DNS Cache Settings in Uptime Kuma

There was a lot of speculation about DNS resolution delays and caching, especially with Docker.

@CommanderStorm (2023):

> We replaced this with the [NSCD (Name Service Caching Daemon)](https://de.wikipedia.org/wiki/Name_service_caching_daemon). For native installs, you must install this yourself.

My current Kuma settings:

1. ✅ **Enable NSCD** → Enabled  
2. ❌ **(Deprecated) Enable DNS Cache for HTTP(s) monitors** → Disabled  

If others could share their settings too, that might help correlate patterns.  
@jorismak @robsontenorio @sobakintech @leikoilja

---

### 5. Also Occurs in Latest Beta

@leikoilja reported the issue even with `2.0.0-beta.3`:

> Running on Docker (Ubuntu 24), same issue with v1 and 2.0.0-beta.3.  
> Makes it hard to trust Uptime Kuma for infra reliability.

---

### 6. Reverse Proxy (Nginx) - till now not mentioned in this thread

This hasn’t been discussed yet, but I wonder if the **reverse proxy (Nginx with SSL)** setup could be involved.

- I followed [the official reverse proxy guide](https://github.com/louislam/uptime-kuma/wiki/Reverse-Proxy)
- No custom tuning - just a default Nginx config
- Could this layer **add latency**, especially when Cloudflare is also in the mix?

I can imagine Uptime Kuma is waiting for a response from the origin server or Cloudflare, but due to additional latency introduced by Nginx, it's hitting the timeout. But 48 seconds in total is still a lot of time.

---

### 7. Why 48s Timeouts?

I don’t fully understand how these layers - Docker, DNS, Cloudflare, Reverse Proxy - **together** could lead to hitting **48 seconds** regularly, but:

- Maybe each layer contributes small delays  
- Cloudflare’s edge behavior (e.g., TLS handshake, rate limits) might amplify it  
- Reverse proxy might be adding overhead or buffer delays  

Raising timeout values to 60s isn’t a solution and also not possible with Uptime Kuma where the default and maximum is 48 seconds - **Uptime Robot**, for reference, gives up after just 30 seconds and never reported those false downtimes like Uptime Kuma does for years.

---

### 8. Final Notes and Attempted Workaround

For now I set all monitors to `Retry = 1`.  
It feels like a workaround and it feels like cheating, but I’ve spent **countless hours** on this over the past year. I’m just a WordPress guy who wanted to self-host uptime monitoring. If I had known earlier, I might have chosen a different solution. But here I am, still hoping to help get to the bottom of this.

**Comment by sobakintech** [comment]
The timeout issue is indeed caused by Cloudflare. I've just setup Uptime Kuma on Oracle Cloud, and I don't get any timeouts anymore. Also I copied the data folder from my home server so it's the exact same config.

**Comment by jorismak** [comment]
I'll keep searching with the non-cloudflare users then 😉

**Comment by raabmar** [comment]
> The timeout issue is indeed caused by Cloudflare. I've just setup Uptime Kuma on Oracle Cloud, and I don't get any timeouts anymore. Also I copied the data folder from my home server so it's the exact same config.

I am curious about the connection here - could you clarify please?

You set up Uptime Kuma on Oracle Cloud and do not get any timeouts anymore → how is it related to Cloudflare? Till now when Cloudflare was mentioned it was about individual websites being behind Cloudflare, thus it means Cloudflare adds probably an additional security layer and also additional routing points and delays to this all which could be therefore part of the reasons for the timeouts

What was your previous setup? Maybe you run Uptime Kuma in a docker container previously, and now not anymore.

**Comment by sobakintech** [comment]
> > The timeout issue is indeed caused by Cloudflare. I've just setup Uptime Kuma on Oracle Cloud, and I don't get any timeouts anymore. Also I copied the data folder from my home server so it's the exact same config.
> 
> I am curious about the connection here - could you clarify please?
> 
> You set up Uptime Kuma on Oracle Cloud and do not get any timeouts anymore → how is it related to Cloudflare? Till now when Cloudflare was mentioned it was about individual websites being behind Cloudflare, thus it means Cloudflare adds probably an additional security layer and also additional routing points and delays to this all which could be therefore part of the reasons for the timeouts
> 
> What was your previous setup? Maybe you run Uptime Kuma in a docker container previously, and now not anymore.

I meant there are no timeouts on websites behind cloudflare. Also Cloudflare protected websites were the only one that were getting timeouts for me. I have the exact same setup - I copied the Uptime Kuma's data folder to the Oracle server.

**Comment by Coysh** [comment]
Just coming from #5905 - I've got a very similar issue. All of my sites are behind Cloudflare. Has seemed to get worse in the last week to a point I'm having to turn off notifications.

This screenshot shows that it is affecting most of my sites. All of these sites are behind Cloudflare - a lot of them are low traffic and sit on the same server.

![Image](https://github.com/user-attachments/assets/df818816-291f-4378-b779-51a2330f66cb)

**Comment by raabmar** [comment]
> I meant there are no timeouts on websites behind cloudflare. Also Cloudflare protected websites were the only one that were getting timeouts for me. I have the exact same setup - I copied the Uptime Kuma's data folder to the Oracle server.

Sorry @sobakintech it still is not clear. To me it sounds like two different things in one paragraph

**Comment by raabmar** [comment]
> Just coming from [#5905](https://github.com/louislam/uptime-kuma/issues/5905) - I've got a very similar issue. All of my sites are behind Cloudflare. Has seemed to get worse in the last week to a point I'm having to turn off notifications. [...]

You could check the entire thread in here or the summary here to get a better understanding of this unresolvable situation we are all facing https://github.com/louislam/uptime-kuma/issues/275#issuecomment-2954334021

Overall it seems it's a problem that comes from a combination of several factors to an unknown extent. The problem is that no person seems to be responsible and/or knowledgeable enough to be able to make progress here. It's definitely frustrating as is. I assume from what I know it's a combination of Docker setup + sites that are monitored behind Cloudflare. It seems this combination leads to the most issues with timeouts. This does not mean that another combination does not make any problems, it's just the one that I have seen most often in here

**Comment by sobakintech** [comment]
> > I meant there are no timeouts on websites behind cloudflare. Also Cloudflare protected websites were the only one that were getting timeouts for me. I have the exact same setup - I copied the Uptime Kuma's data folder to the Oracle server.
> 
> Sorry @sobakintech it still is not clear. To me it sounds like two different things in one paragraph
> 

The timeouts were happening for Cloudflare protected websites, and after switching to Oracle they're completely gone, so seems like Cloudflare is suspicious about IP's with low reputation. 

I copied the data folder from my home server to the Oracle one, so it's the exact same configuration. Is that clear now?

**Comment by elton-bt** [comment]
In my scenario, I have uptime kuma running on docker in a Ubuntu VM on
Oracle Cloud. The monitored sites are behind a reverse proxy with nginx
hosted in my homelab. I don't use cloudflare and the timeout error messages
were so many that I had to disable notifications as well.




Em qui, 12 de jun de 2025 07:12, SobakinTech ***@***.***>
escreveu:

**Comment by raabmar** [comment]
> > > I meant there are no timeouts on websites behind cloudflare. Also Cloudflare protected websites were the only one that were getting timeouts for me. I have the exact same setup - I copied the Uptime Kuma's data folder to the Oracle server.
> > 
> > 
> > Sorry [@sobakintech](https://github.com/sobakintech) it still is not clear. To me it sounds like two different things in one paragraph
> 
> The timeouts were happening for Cloudflare protected websites, and after switching to Oracle they're completely gone, so seems like Cloudflare is suspicious about IP's with low reputation.
> 
> I copied the data folder from my home server to the Oracle one, so it's the exact same configuration. Is that clear now?

OK, so it means:
before → no Docker setup
after → no Docker setup but different IP address where Uptime Kuma monitoring runs from

before → target sites (the one you monitor) were behind Cloudflare
after → same, no difference

You did not change anything on the individual sites.
You did not change anything on the individual Cloudflare setup for each site.
I assume you do not have any IP access rules in Cloudflare (e.g. IP whitelisting) as probably you would have mentioned them.

So in the end your conclusion in your scenario is that the IP address changed and that made the difference. Maybe besides the IP address the location changed and there is a different routing that makes the difference.

Difficult to see the big picture of this entire case

**Comment by mattewid** [comment]
Just to report, I also have this problem **timeout by AbortSignal (48s)** very often.

And I use Cloudflare on my site. And when I pause Cloudflare, this problem simply disappears. So I believe the problem can only be Cloudflare.

I've tried things like creating a rule to ignore Cloudflare security when a specific Header is passed in the uptime kuma request, but the error still persists.

**Comment by jorismak** [comment]
Clearly multiple things going on.

It appears that Cloudflare learns to throw away the 'pings' that Uptime
Kuma makes, resulting in timeouts in your tests. If this is a mistake, or a
'feature', we don't know. But people are clearly having issues with
Cloudflare on their sites that go away if Cloudflare is bypassed (or their
proxy is disabled? What do you mean if you 'pause' Cloudflare).

But it also appears (lots of) timeouts can appear without Cloudflare in the
picture at all. For me this seems performance related. A small VPS with
hundreds of tests had lots of timeouts. Setting the retry to 3 and the
tests to every hour means I basically have zero timeouts. The hope is/was
that the 2.x release will be more efficient. UptimeKuma eats basically all
the RAM my small VPS has available, while a complete Gitea deploy with a
Pipeline runner takes up less ram :).

**Comment by mattewid** [comment]
> What do you mean if you 'pause' Cloudflare)

When accessing the main page of a domain, at the bottom of the page there is the option "Pause Cloudflare"

![Image](https://github.com/user-attachments/assets/006a9509-8d66-4626-8a1c-39ff2a3dc822)

**Comment by Coysh** [comment]
Maybe unrelated - but looks like Uptime Robot is also having a similar issue: https://status.uptimerobot.com/history

"Temporary Monitoring Degradation in EU Region
Some users may experience false-positive alerts due to upstream provider restrictions affecting our monitoring servers in the EU. We’re actively working on a solution to restore full service."

**Comment by raabmar** [comment]
> Maybe unrelated - but looks like Uptime Robot is also having a similar issue: https://status.uptimerobot.com/history
> 
> "Temporary Monitoring Degradation in EU Region Some users may experience false-positive alerts due to upstream provider restrictions affecting our monitoring servers in the EU. We’re actively working on a solution to restore full service."

Yes it's unrelated. I am using Uptime Robot for many years and that's the very first time today.

**Comment by mattewid** [comment]
I realized that the problem is not really with Cloudflare. It's with Uptime Kuma.

Because I self-hosted https://github.com/statping-ng/statping-ng and I don't have this problem with false alerts and I have 100 percent uptime.

**Comment by leikoilja** [comment]
@mattewid , how is your experience with statping-ng? I started migrating to https://github.com/bluewave-labs/Checkmate, which is also quite popular and things are stable on that. the only things i m missing in Checkmate for now are: better ntfy support and configuration via files

**Comment by mattewid** [comment]
> [@mattewid](https://github.com/mattewid) , how is your experience with statping-ng? I started migrating to https://github.com/bluewave-labs/Checkmate, which is also quite popular and things are stable on that. the only things i m missing in Checkmate for now are: better ntfy support and configuration via files

It works fine. The only bad part is that it is no longer being maintained. I'm only using it because I couldn't find any better options.

**Comment by mattewid** [comment]
> [@mattewid](https://github.com/mattewid) , how is your experience with statping-ng? I started migrating to https://github.com/bluewave-labs/Checkmate, which is also quite popular and things are stable on that. the only things i m missing in Checkmate for now are: better ntfy support and configuration via files

I didn't know about this checkmate. It seems good, I'll test it later.

**Comment by leikoilja** [comment]
> > [@mattewid](https://github.com/mattewid) , how is your experience with statping-ng? I started migrating to https://github.com/bluewave-labs/Checkmate, which is also quite popular and things are stable on that. the only things i m missing in Checkmate for now are: better ntfy support and configuration via files
> 
> I didn't know about this checkmate. It seems good, I'll test it later.

checkmate is nice but a bit immature for me at this point, so i m waiting until it gets more traction for a full scale migration

**Comment by steveiliop56** [comment]
I have also faced the timeout issue so let me add my own experience to this issue. I usually set all my monitors to 1 retry just to avoid any DNS issues or false positives since uptime kuma is running with docker on a travel server. I had two GitHub pages websites with the domains in Cloudflare. In Cloudflare I only have an A record pointing to the GitHub IP with proxy disabled and the default TTL. I would keep getting timeouts on one of the websites while the other one had no issues. Turns out I had forgotten to set the retries on the website that was getting the timeouts. Keep in mind that nothing was down when uptime kuma reported it. So it seems like the issue is related to uptime kuma because the moment it retries, it succeeds.

**Comment by ManuelDrescher** [comment]
_(Excuse my english, it is not my native language)_

Same problem here with LXC (Linux container) running on Proxmox. The host is a vServer with dedicated ressources.
Uptime Kuma is running within a LXC and monitors my other services also running as a LXC (same internal network).

I'm currently only monitoring 5 services and getting 30+ "48000ms exceeded" errors per day, but all of them are working.
I haven't found any solution yet.

**Comment by Morfilio** [comment]
@ManuelDrescher 

I have exactly the same problem on same environment, only that i'm monitoring external services. I'm migrating to a full Ubuntu VM right now to test if it's ok.

**Comment by Giuseppe-sciarra** [comment]
I have the same problem, I have some websites hosted on my server, and uptime Kuma on oracle cloud, no Cloudflare, I have configured 5 retries every 180s and seems to solve, for me

**Comment by DarkLiKally** [comment]
I'll also add my experience:

Uptime Kuma is currently running as docker, with tag latest
I'm monitoring currently 12 internal services.
All are monitoring https hosts
All can be resolved through the same identical DNS Server
Tested monitoring by internal and external access.
I'm not using Cloudflare

For 9 out of the 12 services it works fine, never a single issue monitoring those.
For 3 of the services it constantly shows the same picture as in the issue creators post, toggling between 200 OK and timeout.

I already tried configuring some static files on those services to monitor, the timeout options, using chromium check instead of normal http(s) check, retries and so on... Nothing helps resolving the issues for those 3 services.
In my Browsers or on any host I can resolve the monitored urls without any issues.


What I can observe is:
Setting retries to 0 => I get the alternating pattern between 200 OK and timeout
Setting retries to 1 => I get constantly timeout

Another side note: 2 of the failing monitors are hosted on the same machine as 3 of the always successful ones - so its definitely not machine or DNS related. I also tried monitoring by ip and port diectly to avoid any DNS issue => same behaviour

My question is - does Uptime Kuma expect some kind of special thing happening on the HTTP Server side? Expecting a special header or something that those 3 services do not deliver?

One weird observation: for one of the 3 services that fail, the failing started just 2 days ago, before it was fine for about a week.


UPDATE:

I enabled the request logging on my reverse proxy, and noticed that no requests to the monitored failing domains are coming through when only uptime kuma is monitoring it (no user active).
I did the same on the DNS Server, here I can see the requests coming in every ~ 1 minute as configured in uptime kuma.
DNS Server logs say, the requests are answeres correctly.
So my assumption is something seems to go wrong on the uptime-kuma side between the DNS resolution and the executing the requests to the HTTP endpoints since the reverse proxy is never receiving requests to the failing domains.
As mentioned above there are other applications hosted on the same host, behind the same reverse proxy, DNS and so on, and those can be monitored totally fine. here I can also observe the incoming requests on the reverse proxy logs.

Another thing I found now:
I can sometimes observe the occurense of  the following entry in the uptime kuma UI for some of the WORKING hosts.
```
Inaktiv | 2025-08-25 09:36:46 | getaddrinfo EAI_AGAIN masked.domain.example.com
```
but it directly retries and works again.

**Comment by adamalfath** [comment]
Is there a way to suppress the timeout notification?
For my purpose. I could “ignore” the timeout status, but other status update are still helpful.

Thanks.

**Comment by SeanKilleen** [comment]
(Before I begin: love the project, love the work you've done on it, and I appreciate you! Just sponsored via GitHub.)

Hi all,

Wanted to add some additional data points and a theory in case it's helpful. We're experiencing the same issue and it's made uptime-kuma unusable for us as the number of false positives is overwhelming without pretty consistently restarting the containers.

We're running uptime-kuma as an Azure Container App (we have one instance for our pre-prod environment and one for prod). Both container apps experience this issue. Over the last 7 days, the containers use max 8% of their allotted RAM and max 3.5% of their allotted CPU. They are both backed by independent MySQL databases, hosted in Azure, that also show CPU at 5% and RAM at 25%.

We have roughly 58 monitors in prod. I have correlation from our logs to indicate that these are false positives from uptime-kuma.

A few things I've seem to notice, but can't prove a correlation to yet

* The number of alerts seem to rise over time until they get back enough that I restart the uptime-kuma container
* For some time after that restart, false positives cease. They then begin ticking back up again.
* I have added 1 retry for every item, but that hasn't prevented the uptick in false positives.
* There seem to be a great deal of queries against the DB -- here's the last 24 hours

<img width="2238" height="769" alt="Image" src="https://github.com/user-attachments/assets/c9da0c9f-14da-4e9d-9f16-c270e5ea4b21" />

* I'm not sure if connections are pooled, but there appear to be large spikes of connections -- here's over the last 7 days:

<img width="2221" height="832" alt="Image" src="https://github.com/user-attachments/assets/154a890c-a6f3-4a35-8625-a2e7d39cb799" />


Between the large spikes of connections and the fact that this problem appears to be gradually introduced after every restart, a theory I have is that something around database connection pooling might not be handled / released properly. It seems like a leak of some kind. I don't know enough about the stack / codebase to verify that.

Happy to provide other stats or dig into this in any way that might be helpful.

**Comment by SeanKilleen** [comment]
I would normally bristle at this myself, but armed with a little bit of knowledge of what we were experiencing, I forked the codebase and did some analysis assisted by Claude.

It appears (from what I could verify of Claude's observations ) that:

In: `server/model/monitor.js:522-533` every time a monitor performs an HTTP check, it creates brand new HTTP/HTTPS agents:
```
if (!options.httpAgent) {
    options.httpAgent = new http.Agent(httpAgentOptions);
}
```

```
if (!options.httpsAgent) {
    let jar = new CookieJar();
    let httpsCookieAgentOptions = {
        ...httpsAgentOptions,
        cookies: { jar }
    };
    options.httpsAgent = new HttpsCookieAgent(httpsCookieAgentOptions);
}
```

These agents are never destroyed after the request completes. Over time, agents accumulate in memory, and each agent maintains socket pools and internal state. Eventually, the system runs out of resources, causing the 48000ms timeout errors

Additional Possible Issues Found:

* NTLM KeepAlive: At `server/model/monitor.js:1047` there's a keepAlive = true set for NTLM authentication that's never cleaned up
* No cleanup in `stop()`: The monitor's `stop()` method only clears the timeout but doesn't destroy agents
* Proxy agents: Similar issue in `server/proxy.js:94-160` where proxy agents are created but never destroyed.

___

Based on this, I'm going to attempt to do a PR, in the hopes that it will get some more thorough review and possibly resolve the isuse.

**Comment by steveiliop56** [comment]
This wouldn't make sense. JavaScript has a garbage collector which should cleanup the clients.

**Comment by SeanKilleen** [comment]
> This wouldn't make sense. JavaScript has a garbage collector which should cleanup the clients.

I would typically be wary of this as well, if it didn't exactly match my experience of monitors timing out more and more due to http timeouts, and where the only resolution seems to be to restart the container.

If folks are willing to provide a little guidance, I'm happy to build a version of the container based on this PR, put it in our container registry, and use it in our production use case. That might be the ultimate test of whether it resolves things.

**Comment by steveiliop56** [comment]
Even in that case the failure rate should be linear. E.g. you start with a few failures and reach 100% failure rate at one point (assuming resources are the issue). The actual test here would be to use some sort of automated tool (perhaps even by using the uptime kuma API) to generate thousands of monitors and see when things start to fail. I personally have sort of resolved the issue by using 2-3 retries, definitely not the best solution but seems to be working well enough.

**Comment by SeanKilleen** [comment]
@steveiliop56 

>  The actual test here would be to use some sort of automated tool (perhaps even by using the uptime kuma API) to generate thousands of monitors and see when things start to fail.

I welcome your contribution to my PR to do that! Or a separate PR. Or a walkthrough on how to do that here. Or an example of where you've done so.

> I personally have sort of resolved the issue by using 2-3 retries, definitely not the best solution but seems to be working well enough

Increasing the number of retries is not a viable solution for our monitoring use case.

---

Rather than discuss theoretical fixes / reasons why a fix couldn't work, I'd prefer to wait for a maintainer to weigh in on this issue or on the PR. In he interim, I may build and deploy my fork to see if it has any practical implications. Since this issue has been reported by folks since 2021, I think the solution will end up being non-obvious.

**Comment by louislam** [comment]
@SeanKilleen has created a pull request #6380 to fix this issue. I don't have this issue personally, I can't verify it, so I need you guys to test it.



Command:

```
npx kuma-pr SeanKilleen:fix-db-issue
```

or this if you don't have Node.js

```
docker run --rm -it -p 3000:3000 -p 3001:3001 --pull always -e 'UPTIME_KUMA_GH_REPO=SeanKilleen:fix-db-issue' louislam/uptime-kuma:pr-test2
```

https://github.com/louislam/uptime-kuma/wiki/Test-Pull-Requests

**Comment by etsiej** [comment]
Having the same issue as @SeanKilleen trying his version of monitor.js on my Kuma container to see if it reduces or stops the timeouts.

**Comment by vedmaka** [comment]
We are running around 50 HTTP monitors, with every monitor configured to perform the check every 60 seconds. From time to time, we experience timeout issues

The DNS check did not help, and it seems the Uptime Kuma application just hangs the connections at some point, and this causes a cascading failure for the rest of the monitors configured, as the timeout is usually thrown for one monitor, and then a bunch of others follow until it gets back to normal

The pattern reported by the Uptime graph is usually the same - there is a normal response time on the graph as expected. Then there is a timeout reported. Then the service is reported as back with a greatly increased response time, and then subsequent checks report response time as usual

All the monitors that catch that timeout issue show the same pattern

While this happens, the resource consumption of the container is quite subtle, so this is unlikely to be related to resource exhaustion

**Comment by bangnokia** [comment]
after 5 years this still exists 😂

<img width="1150" height="1023" alt="Image" src="https://github.com/user-attachments/assets/86e85f54-92a5-4c3a-8833-e9946a6e8251" />

**Comment by SeanKilleen** [comment]
Hi @bangnokia, please note that I've proposed a fix and it's available to test -- please see the last few comments if you'd like to help us confirm whether the fix resolves the issues.

**Comment by bangnokia** [comment]
> Hi [@bangnokia](https://github.com/bangnokia), please note that I've proposed a fix and it's available to test -- please see the last few comments if you'd like to help us confirm whether the fix resolves the issues.

that's great, let me try

**Comment by carlosflorencio** [comment]
Just to add to this thread because I had this exact issue, what improved on my side was to enable Optimistic cache in Adguard Home (my local dns server). 

I noticed some dns queries taking up to 70ms to resolve (using an upstream via DoH) and by enabling this setting all queries started being much faster and the amount of timeouts reduced a lot.

**Comment by etsiej** [comment]
I've been running the updated Kuma with the new monitor.js for a few days in Azure without much success. 

<img width="1065" height="1008" alt="Image" src="https://github.com/user-attachments/assets/2cfac902-f248-403a-8fe8-91d59adecb63" />

**Comment by SeanKilleen** [comment]
Unfortunately, I have to admit defeat on my PR. While it extended the timeframe between symptoms occurring, it did not resolve them.

Unfortunately, we had a real need to stop these false positives and my attempted solution didn't work. I ended up having to migrate us away to a different solution.

I closed my PR, but someone else can feel free to resurrect it and run with it if there are parts that are helpful. I'm sorry we weren't able to figure it out!

**Comment by etsiej** [comment]
Still a mystery to me as well. unfortunately i don't get the time this side of Christmas to dig any deeper.
@SeanKilleen what other open source solution are you trialing out of interest?

**Comment by SeanKilleen** [comment]
> Still a mystery to me as well. unfortunately i don't get the time this side of Christmas to dig any deeper. [@SeanKilleen](https://github.com/SeanKilleen) what other open source solution are you trialing out of interest?

Out of respect for this project and its maintainer, I'm not going to advertise that here. 👍OSS maintainers put a lot of time and sweat into these projects and I don't want to be the reason people abandon the solution. I did some research, selected an alternative that I knew to be Docker compatible, and converted it when I had a chunk of time to do so. I think I would have still preferred the uptime-kuma UI etc. if I could have gotten it working, and I certainly hope someone is able to get to the resolution!

**Comment by CommanderStorm** [comment]
Hey, sharing what worked for you is totally fine by me. In the end, we won't be able to make everyone happy.

Because this issue is so hard to reproduce it is particularly hard to fix.
If the PR you opened really moves us closer to fixing this, that would be awesome. I don't fully get why that approach would work though, so further debugging would be required.
This might be a bug in node instead, or in how we do timeouts.

**Comment by SeanKilleen** [comment]
Wanted to let this group know that I'm exploring a different theory now.

A few days after I deployed the other solution I mentioned, similar symptoms started happening. Random timeouts with similar errors, on I totally different tool with a totally different tech stack.

I did a little research and my new theory is that this might have been related to SNAT exhaustion. I didn't realize that Azure container apps pools SNAT. If I'm being honest, I wasn't previously very familiar with SNAT at all.

I'm testing this theory by migrating the current tool to azure container instances, which does not pool/share SNAT as far as I'm aware. So far so good. I'm going to let it run for a week and see if that fixes the symptoms. If so, for the sake of the community, I'll migrate the uptime Kuma instance I currently have shut down, and see if that solves the symptoms. If so, I think we may have a strong contender for what causes the issue (even if there aren't a ton of options for uptime-kuma to address it).

Will keep you all posted here.

**Comment by carlosflorencio** [comment]
Just to add more ideas to this thread, trying a different http client like [undici](https://github.com/nodejs/undici) could be beneficial since it handles tcp sockets differently. 

The dns resolution strategy may also cause this issue, there are some libs adding async dns resolution to `axios` like this one: https://github.com/tcollinsworth/axios-cached-dns-resolve.

Updating `axios` to a more recent version could also fix some issues?

**Comment by SeanKilleen** [comment]
An update -- about 6 days later, no recurrence of the symptom since we moved our alternate setup to Azure Container Instances from Azure Container Apps. Will probably let it run for another week.

**Comment by etsiej** [comment]
SNAT exhaustion might actually explain the issues I was seeing, too. Although my Kuma instance ran as a container on a dedicated server (rather than an App Service), 

I was monitoring an msappproxy.net endpoint and seeing the exact same symptoms mentioned in this thread. I'm fairly certain that traffic is subject to SNAT translation as well.

There’s a great write-up on how this works [here](https://4lowtherabbit.github.io/blogs/2019/10/SNAT/#symptom-when-snat-port-exhausted).

**Comment by SeanKilleen** [comment]
Update -- I think that it was indeed a SNAT issue. I've had zero issues since moving azure services. Considering this mystery solved.

Thanks to everyone! Feel free to close this.

**Comment by m-stefanski** [comment]
We are experiencing the same log abnormalities running in podman on RHEL VMs. 

```
2026-02-03T14:34:07+01:00 [MONITOR] WARN: Monitor #1 'XXX QA': Failing: timeout of 48000ms exceeded | Interval: 60 seconds | Type: http | Down Count: 0 | Resend Interval: 0
2026-02-03T14:34:07+01:00 [MONITOR] WARN: Monitor #2 'XXX UAT': Failing: timeout of 48000ms exceeded | Interval: 60 seconds | Type: http | Down Count: 0 | Resend Interval: 0
2026-02-03T14:34:08+01:00 [MONITOR] WARN: Monitor #5 'YYY QA': Failing: timeout of 48000ms exceeded | Interval: 60 seconds | Type: http | Down Count: 0 | Resend Interval: 0
2026-02-03T14:34:09+01:00 [MONITOR] WARN: Monitor #6 'YYY UAT': Failing: timeout of 48000ms exceeded | Interval: 60 seconds | Type: http | Down Count: 0 | Resend Interval: 0
```

Applications are reachable all the time

**Comment by Jojonintendo** [comment]
I run all my services via Podman containers. Most of them run with Pasta networking.

Configuring Caddy to use "host" networking completely fixed this issue on my end. It's been a month since I made the switch, not a single timeout.

**Comment by pankajmyt** [comment]
been using uptime, love the UI, the configurable alerts, ability to monitor db and services. The alerts arent stable, every couple of day we get a series of down, then up alerts. Service wouldnt be going through downtime

**Comment by Thanawat1997** [comment]
I have the same issue in my environment, like this:
**Uptime Kuma Host:** Self-hosted on a VM on Azure
**Application:** Next.js version 15.3.8 that runs on Docker (node:24.11-bullseye) in Azure Container Apps

<img width="1429" height="1423" alt="Image" src="https://github.com/user-attachments/assets/ae81b8d5-22a5-47aa-bf5b-1fa2ca8e60f7" />

**Comment by steveiliop56** [comment]
I believe the issue is not related to uptime-kuma at all.

As with everyone here, I also started exploring alternatives and tried Gatus which is written in Go (more networking focused language, native binary - essentially does things differently), deployed in Docker in the same way I deployed uptime-kuma and next thing you know client.Timeout was reached.

This issue probably lives somewhere in the operating system itself (or even Docker) and not with the monitoring implementations themselves.

If anyone is interested in helping I believe a good test would be to run 4 total monitors. One simple node script running in Docker hitting an endpoint every 60 seconds and writing the status to a log file, same script running inside Docker and another 2 scripts but in Go this time. This way we will know if the issue is related to node (probably isn't), Docker (most likely I believe) or the operating system itself.

**Comment by SeanKilleen** [comment]
@steveiliop56 see my comment from a few back. It's definitely not uptime-kuma's fault. It's SNAT exhaustion. I hit this with both uptime-kuma and Gatus. Both were fixed when I moved Azure services to one with less constrained SNAT.

If my monitors hit the tipping point again I'll either be backing things off or paying a small bit (from a business perspective) for a managed NAT gateway, which wouldn't have that issue at all.

**Comment by steveiliop56** [comment]
@SeanKilleen that only seems to apply to Azure not general Docker deployments.

**Comment by SanderFlooris** [comment]
We recently moved our Uptime Kuma from a Hetzner VPS to a DigitalOcean Droplet so we can also monitor hosting provider outages, and we noticed that it seemed to no longer get those random timeouts.

Both are running via Docker using the same configuration _(as far as we know atleast)_

**Edit:** Clarifying that I'm talking about the Docker engine configuration, the Docker Compose configuration is exactly the same. Not sure if certain hosting providers customize the Docker engine configuration by default.

**Comment by steveiliop56** [comment]
@louislam I think it's safe to close this issue to be honest. Just confirmed it's not really related to uptime-kuma but rather it's some underlying docker or operating system bug. Deployed Gatus sometime ago which runs natively-ish (ish due to the Go runtime) since it's Go and guess what, same issue:

![Screenshot](https://github.com/user-attachments/assets/2e403c76-2f6e-4135-b700-8b77041bdb63)

*This is a website hosted via GitHub pages, 0 downtime, same 48000 timeout*

My *new* theory would be resource exhaustion. Maybe running these in parallel exhausts some OS limit and triggers the timeout? Would it make sense to add some sort of jitter between the uptime checks instead of running them simultaneously?

**Comment by CommanderStorm** [comment]
> [@steveiliop56](https://github.com/steveiliop56) see my comment from a few back. **It's definitely not uptime-kuma's fault. It's SNAT exhaustion.** I hit this with both uptime-kuma and Gatus. Both were fixed when I moved Azure services to one with less constrained SNAT.
> 
> If my monitors hit the tipping point again I'll either be backing things off or paying a small bit (from a business perspective) for a managed NAT gateway, which wouldn't have that issue at all.

@m-stefanski the pinned message must have gotten too far up the thread again for you to see it.
I wonder if we can somehow detect this and raise a better error message.
I have not thought about this problem or done research into this.
It is defintively a shortcoming where we could do better, but I don't know how.

**Comment by louislam** [comment]
@etsiej @SeanKilleen @steveiliop56 Thank you guys, I think we finally solved this long time mystery.

I think it is a big finding. Imagine, if you deploy other serious applications on such providers, which your applications are highly depending on HTTP requests, you will likely run into the same situation.

One last thing, I would like to consolidate providers that are having SNAT exhaustion issue:
- Azure Container Apps
- Azure App Service
- Azure Services
- Azure Kubernetes Service (AKS)
- Google Kubernetes Engine
- (More?)

P.S. Azure Container Instances is not affected

**Comment by mattewid** [comment]
I'm sorry to say, but I believe the problem hasn't been resolved. In the past, I hosted Uptime Kuma on a local server without any kind of blocking, and this problem still occurred. After migrating to https://github.com/bluewave-labs/checkmate, I haven't had this problem anymore. I've been running Checkmate for months.

**Comment by steveiliop56** [comment]
@CommanderStorm yes, could be SNAT exhaustion but this only seems to apply to cloud providers and not self-hosted instances.

@mattewid it's not resolved indeed but it's not related to uptime-kuma either. Did you deploy checkmate on the exact same server? Was it native or Docker?

**Comment by mattewid** [comment]
> [@CommanderStorm](https://github.com/CommanderStorm) yes, could be SNAT exhaustion but this only seems to apply to cloud providers and not self-hosted instances.
> 
> [@mattewid](https://github.com/mattewid) it's not resolved indeed but it's not related to uptime-kuma either. Did you deploy checkmate on the exact same server? Was it native or Docker?

The installation was done on the same server and using Docker.

**Comment by sofia-fernandez-six** [comment]
> [@etsiej](https://github.com/etsiej) [@SeanKilleen](https://github.com/SeanKilleen) [@steveiliop56](https://github.com/steveiliop56) Thank you guys, I think we finally solved this long time mystery.
> 
> I think it is a big finding. Imagine, if you deploy other serious applications on such providers, which your applications are highly depending on HTTP requests, you will likely run into the same situation.
> 
> One last thing, I would like to consolidate providers that are having SNAT exhaustion issue:
> 
> * Azure Container Apps
> * Azure App Service
> * Azure Services
> * Azure Kubernetes Service (AKS)
> * Google Kubernetes Engine
> * (More?)
> 
> P.S. Azure Container Instances is not affected

For your last comment, I think we are experiencing similar issue, but we started since we upgraded to version 2.2.1 and migrated from container instance to container apps in Azure, our problem is sporadically we receive a timeout to a 
[36m2026-04-23T10:29:56+02:00[0m [[38;5;119mMONITOR[0m] [33mWARN:[0m Monitor #23 '[A0355-qa-external-ch] apip-catalog-external-v1': Pending: timeout of 60000ms exceeded | Max retries: 3 | Retry: 1 | Retry Interval: 60 seconds | Type: http

Uptime is hosted as Azure Container Apps and the monitor is a developer portal of Kong Konnect Platform, we checked with Kong and no issues from their side

**Comment by Buco7854** [comment]
Hi, I run Uptime Kuma in Docker on a cheap 1GB RAM, 1 vCPU Debian 13 IONOS VPS and had the same issue. It seemed to only impact sites behind Cloudflare.
I tried disabling Browser Integrity Check, and the SNAT exhaustion theory but that wasn't it.
The fix for me was to add this to the Uptime Kuma service in my docker compose:

```yaml
sysctls:
  - net.ipv6.conf.all.disable_ipv6=1
  - net.ipv6.conf.default.disable_ipv6=1
```

The theory (Got to it with Claude) is that Cloudflare proxied sites return both IPv4 (A) and IPv6 (AAAA) records. Docker containers don't have working IPv6 by default, so when a check landed on an IPv6 address it hung on the dead route until the timeout hapened, while checks that hit IPv4 were fine (Seems that it rotates between the returned Ips), which is why it looked random and only affected Cloudflare sites. Disabling IPv6 in the container stops the resolver from returning AAAA records, forcing IPv4 and stopping the timeouts.

I can't exactly confirm the theory, but I've had zero issues for 10 hours, after getting them hourly before.

For those who needs to monitor IPv6 only, I guess you'd instead need a working IPv6 address on the VPS and IPv6 configured in Docker.

I'll update if I get a timeout again but if you have no news it means this has fixed it for me.

EDIT: I've been running it for a bit and didn't have the issue for about a day, then I got it 2 or 3 times in the last week. I disabled my fix to test it and got the issue about 4 times in one night, but then nothing. So, I'm afraid I can't be really sure if it definitely did something or if the timing was just a coincidence. If you're experiencing the issue, I guess you can still try it and see if it improves things for you.
